### PR TITLE
Matrix Triangular Solve GPU op

### PR DESCRIPTION
**Pull request by c0g**
Adds a GPU version of the triangular solver op using cuBLAS trsm.

### CODE REVIEW

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc [hunk 1]
```diff
@@ -1,11 +1,8 @@
 /* Copyright 2015 The TensorFlow Authors. All Rights Reserved.
-
 Licensed under the Apache License, Version 2.0 (the "License");
 you may not use this file except in compliance with the License.
 You may obtain a copy of the License at
-
```
Comment (ebrevdo) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: please don't change whitespace in lines you aren't otherwise modifying

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc [hunk 2]
```diff
@@ -24,15 +20,117 @@ limitations under the License.
...
+
+    TensorShapes GetOutputMatrixShapes(
+        const TensorShapes& input_matrix_shapes) const final
+    {
+        return TensorShapes({ TensorShape({ input_matrix_shapes[0].dim_size(1),
+            input_matrix_shapes[1].dim_size(1) }) });
+    }
+
+    int64 GetCostPerUnit(const TensorShapes& input_matrix_shapes) const final
+    {
```
Comment (ebrevdo) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: you probably need to multiply this by Eigen's AddCost / MulCost to get appropriate estimates

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc [hunk 3]
```diff
@@ -24,15 +20,117 @@ limitations under the License.
...
+        double rows = static_cast<double>(input_matrix_shapes[0].dim_size(0));
+        double num_rhss = static_cast<double>(input_matrix_shapes[1].dim_size(1));
+        double cost = rows * rows * num_rhss;
```
Comment (ebrevdo) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: why are you doing this in double precision?

File: tensorflow/python/kernel_tests/matrix_triangular_solve_op_test.py [hunk 4]
```diff
@@ -119,4 +121,4 @@ def testEmpty(self):
 
 
 if __name__ == "__main__":
-  tf.test.main()
+  tf.test.main()
```
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/matrix_triangular_solve_op_test.py]: undo whitespace changes

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc [hunk 5]
```diff
@@ -76,40 +174,89 @@ class MatrixTriangularSolveOp : public LinearAlgebraOp<Scalar> {
...
+    if (adjoint_) {
+      trans = perftools::gputools::blas::Transpose::kTranspose;
     } else {
-      auto triangle = matrix.template triangularView<Eigen::Upper>();
-      if (adjoint_) {
-        output.noalias() = triangle.adjoint().solve(rhs);
-      } else {
-        output.noalias() = triangle.solve(rhs);
-      }
+      trans = perftools::gputools::blas::Transpose::kNoTranspose;
```
Comment (ebrevdo) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: can you use longer variable names instead of 'trans', 'lda', 'ldb', etc?

then in ThenBlasTrsm call, you can do e.g. upper_or_lower /* uplo */, ..., etc.

easier for other users to read.

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc [hunk 6]
```diff
@@ -76,40 +174,89 @@ class MatrixTriangularSolveOp : public LinearAlgebraOp<Scalar> {
...
+    uint64 cublas_m = rhs.cols(); 
+    uint64 cublas_n = rhs.rows(); 
+    bool blas_launch_status =
+        stream
+            ->ThenBlasTrsm(perftools::gputools::blas::Side::kRight, uplo, trans,
+                           perftools::gputools::blas::Diagonal::kNonUnit,
+                           cublas_m, cublas_n, 1.0, matrix_ptr, lda, &out_ptr,
+                           ldb)
+            .ok();
+    // LOG(INFO) << blas_launch_status;
```
Comment (ebrevdo) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: remove commented out code

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc [hunk 7]
```diff
@@ -76,40 +174,89 @@ class MatrixTriangularSolveOp : public LinearAlgebraOp<Scalar> {
...
+    }
+    uint64 lda = matrix.cols();   
+    uint64 ldb = rhs.cols();      
+    uint64 cublas_m = rhs.cols(); 
+    uint64 cublas_n = rhs.rows(); 
+    bool blas_launch_status =
+        stream
+            ->ThenBlasTrsm(perftools::gputools::blas::Side::kRight, uplo, trans,
+                           perftools::gputools::blas::Diagonal::kNonUnit,
+                           cublas_m, cublas_n, 1.0, matrix_ptr, lda, &out_ptr,
```
Comment (ebrevdo) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: I think this needs to be Scalar(1) instead of 1.0  (which is a double)

**Review by ebrevdo — CHANGES_REQUESTED** [review]

**Comment by ebrevdo** [comment]
Thanks for the PR!  A few comments.

File: tensorflow/python/kernel_tests/matrix_triangular_solve_op_test.py [hunk 8]
```diff
@@ -24,15 +24,17 @@
 class MatrixTriangularSolveOpTest(tf.test.TestCase):
```
Comment (ebrevdo) [inline tensorflow/python/kernel_tests/matrix_triangular_solve_op_test.py]: This is not being tested with GPU support.  You need to modify this line:

https://github.com/c0g/tensorflow/blob/e46db8b3e2d1efd5cf10ced68b15cbbd1d58a149/tensorflow/python/kernel_tests/BUILD#L216

from tf_py_test to cuda_py_test.

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc (hunk 3 above)
Comment (c0g) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: This is the original tensorflow op code (see above). Possibly because double can represent larger numbers than int? I doubt anyone will try a matrix multiply large enough to overflow a 64 bit int though. Happy to change to int64.

**Comment by ebrevdo** [comment]
Ah in that case let's leave it as double. You're right.

**Comment by drpngx** [comment]
Do we still need changes?

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by c0g** [comment]
I have not yet made the requested changes.

**Comment by c0g** [comment]
@ebrevdo : I reverted to the original file to fix white space then made your other requested changes.

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc (hunk 1 above)
Comment (drpngx) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: Please undo these changes.

**Comment by drpngx** [comment]
I don't see the revert. Please push?

**Comment by c0g** [comment]
@drpngx you're right I have no idea what happened to them. The changes should now be up. Sorry for the lag/commit spam.

**Comment by drpngx** [comment]
Jenkins, test this please.

**Comment by alexggmatthews** [comment]
Seems to be passing now, only looks to be waiting on verification of @ebrevdo requested changes.

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc [hunk 9]
```diff
@@ -60,7 +77,9 @@ class MatrixTriangularSolveOp : public LinearAlgebraOp<Scalar> {
   int64 GetCostPerUnit(const TensorShapes& input_matrix_shapes) const final {
     double rows = static_cast<double>(input_matrix_shapes[0].dim_size(0));
     double num_rhss = static_cast<double>(input_matrix_shapes[1].dim_size(1));
-    double cost = rows * rows * num_rhss;
+        double cost = rows * rows * num_rhss *
```
Comment (ebrevdo) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: whitespace?

**Review by ebrevdo — APPROVED** [review]
small nit; otherwise LGTM.

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc (hunk 9 above)
Comment (c0g) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: Sorry! Should be gone now.

File: tensorflow/core/kernels/matrix_triangular_solve_op.cc [hunk 10]
```diff
@@ -25,8 +25,25 @@ limitations under the License.
 #include "tensorflow/core/platform/macros.h"
 #include "tensorflow/core/platform/types.h"
 
+#if GOOGLE_CUDA
+#include "tensorflow/core/platform/stream_executor.h"
+#endif // GOOGLE_CUDA
```
Comment (drpngx) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: Sorry, one last thing: we need an extra space before the //

Comment (c0g) [inline tensorflow/core/kernels/matrix_triangular_solve_op.cc]: done

**Comment by drpngx** [comment]
Jenkins, test this please.
